Thanks for the quick turnaround — the ordering fix is exactly right.

**Prior finding — resolved.** Gather now runs first, `_bulk_clear_entity_store` runs after. Partial vector-store failures leave entity links intact for surviving memories. That was the only blocker from the last round.

**New blocker — CI red on an existing test.**

```
FAILED tests/memory/test_decay_usage_notice.py::test_async_delete_all_decay_usage_runs_after_success
AttributeError: 'AsyncMemory' object has no attribute '_entity_store'
```

The commit message says "Replace getattr guard with direct self._entity_store is not None." In production that's fine — `AsyncMemory.__init__` always sets `self._entity_store = None` (line 1978). But `make_async_memory()` in that test file constructs via `AsyncMemory.__new__(AsyncMemory)`, which skips `__init__` entirely, so `_entity_store` is never placed on the object. The `if self._entity_store is not None:` check at `main.py:3261` then raises `AttributeError`.

One-line fix in `tests/memory/test_decay_usage_notice.py`:

```python
def make_async_memory():
    memory = AsyncMemory.__new__(AsyncMemory)
    memory.vector_store = MagicMock()
    memory._delete_memory = AsyncMock()
    memory._entity_store = None   # add this — mirrors AsyncMemory.__init__ line 1978
    return memory
```

The production logic in `delete_all()` is correct; the test helper just needs to reflect the real instance state. Once CI is green, this is good to go.

**Evidence**
- CI: FAIL — Python SDK / build_mem0 (3.11), job 81910849641 — 1 failed, 1451 passed
- Failure: `test_async_delete_all_decay_usage_runs_after_success` — `AttributeError: 'AsyncMemory' object has no attribute '_entity_store'`
- Ordering verified in diff: gather at `main.py:3257–3260`, entity clear at `main.py:3262–3263` — correct

--- a/mem0/memory/main.py
+++ b/mem0/memory/main.py
@@ -3252,14 +3252,14 @@ class AsyncMemory(MemoryBase):
         capture_event("mem0.delete_all", self, {"keys": keys, "encoded_ids": encoded_ids, "sync_type": "async"})
         memories = await asyncio.to_thread(self.vector_store.list, filters=filters)
 
-        if getattr(self, "_entity_store", None) is not None:
-            await self._bulk_clear_entity_store(filters)
-
         delete_tasks = []
         for memory in memories[0]:
             delete_tasks.append(self._delete_memory(memory.id, skip_entity_cleanup=True))
 
         results = await asyncio.gather(*delete_tasks, return_exceptions=True)
+
+        if self._entity_store is not None:
+            await self._bulk_clear_entity_store(filters)
 
         errors = [r for r in results if isinstance(r, BaseException)]
         if errors:

--- a/mem0/memory/main.py
+++ b/mem0/memory/main.py
@@ -2076,6 +2076,27 @@ class AsyncMemory(MemoryBase):
         except Exception as e:
             logger.warning(f"Entity upsert failed for '{entity_text}' (async): {e}")
 
+    async def _bulk_clear_entity_store(self, filters):
+        """Delete all entity records matching the given scope filters.
+
+        Used by delete_all to avoid the race condition that occurs when
+        concurrent _delete_memory coroutines each try to read-modify-write
+        the same entity rows' linked_memory_ids lists.
+        """
+        if self._entity_store is None:
+            return
+        search_filters = {k: v for k, v in filters.items() if k in ("user_id", "agent_id", "run_id") and v}
+        try:
+            listed = await asyncio.to_thread(self.entity_store.list, filters=search_filters, top_k=10000)
+            rows = listed[0] if isinstance(listed, (list, tuple)) and listed and isinstance(listed[0], list) else listed
+            for row in rows or []:
+                try:
+                    await asyncio.to_thread(self.entity_store.delete, vector_id=row.id)
+                except Exception as e:
+                    logger.debug(f"Bulk entity delete failed for id={row.id}: {e}")
+        except Exception as e:
+            logger.warning(f"Bulk entity store cleanup failed: {e}")
+
     async def _remove_memory_from_entity_store(self, memory_id, filters):
         """Async variant of `Memory._remove_memory_from_entity_store`."""
         if self._entity_store is None:
@@ -3231,9 +3252,12 @@ class AsyncMemory(MemoryBase):
         capture_event("mem0.delete_all", self, {"keys": keys, "encoded_ids": encoded_ids, "sync_type": "async"})
         memories = await asyncio.to_thread(self.vector_store.list, filters=filters)
 
+        if self._entity_store is not None:
+            await self._bulk_clear_entity_store(filters)
+
         delete_tasks = []
         for memory in memories[0]:
-            delete_tasks.append(self._delete_memory(memory.id))
+            delete_tasks.append(self._delete_memory(memory.id, skip_entity_cleanup=True))
 
         results = await asyncio.gather(*delete_tasks, return_exceptions=True)
 
@@ -3418,7 +3442,7 @@ class AsyncMemory(MemoryBase):
 
         return memory_id
 
-    async def _delete_memory(self, memory_id, existing_memory=None):
+    async def _delete_memory(self, memory_id, existing_memory=None, skip_entity_cleanup=False):
         logger.info(f"Deleting memory with {memory_id=}")
         if existing_memory is None:
             existing_memory = await asyncio.to_thread(self.vector_store.get, vector_id=memory_id)
@@ -3444,9 +3468,8 @@ class AsyncMemory(MemoryBase):
             is_deleted=1,
         )
 
-        # Entity-store cleanup: strip this memory's id from any entity records
-        # that linked to it. Non-fatal — the helper swallows errors.
-        await self._remove_memory_from_entity_store(memory_id, session_filters)
+        if not skip_entity_cleanup:
+            await self._remove_memory_from_entity_store(memory_id, session_filters)
 
         return memory_id
 

--- a/mem0/memory/main.py
+++ b/mem0/memory/main.py
@@ -3252,7 +3252,7 @@ class AsyncMemory(MemoryBase):
         capture_event("mem0.delete_all", self, {"keys": keys, "encoded_ids": encoded_ids, "sync_type": "async"})
         memories = await asyncio.to_thread(self.vector_store.list, filters=filters)
 
-        if self._entity_store is not None:
+        if getattr(self, "_entity_store", None) is not None:
             await self._bulk_clear_entity_store(filters)
 
         delete_tasks = []

--- a/tests/test_memory.py
+++ b/tests/test_memory.py
@@ -1107,7 +1107,6 @@ class TestHybridSearchWarning:
         assert not any("does not support keyword search" in r.message for r in caplog.records)
 
 
-<<<<<<< HEAD
 class TestPreserveCustomMetadata:
 
     @patch('mem0.utils.factory.EmbedderFactory.create')

--- a/tests/test_memory.py
+++ b/tests/test_memory.py
@@ -1107,6 +1107,7 @@ class TestHybridSearchWarning:
         assert not any("does not support keyword search" in r.message for r in caplog.records)
 
 
+<<<<<<< HEAD
 class TestPreserveCustomMetadata:
 
     @patch('mem0.utils.factory.EmbedderFactory.create')
@@ -1256,3 +1257,52 @@ class TestPreserveCustomMetadata:
         assert payload["source"] == "chat"
         assert payload["data"] == "I love swimming"
         assert payload["user_id"] == "user_1"
+
+
+@pytest.mark.asyncio
+@patch('mem0.utils.factory.EmbedderFactory.create')
+@patch('mem0.utils.factory.VectorStoreFactory.create')
+@patch('mem0.utils.factory.LlmFactory.create')
+@patch('mem0.memory.storage.SQLiteManager')
+async def test_async_delete_all_bulk_clears_entity_store(mock_sqlite, mock_llm_factory, mock_vector_factory, mock_embedder_factory):
+    """
+    Verify that async delete_all bulk-clears entity records before running
+    concurrent memory deletes, preventing the race condition where concurrent
+    _remove_memory_from_entity_store calls read stale snapshots of
+    linked_memory_ids and overwrite each other's writes.
+    """
+    mock_embedder_factory.return_value = MagicMock()
+    mock_llm_factory.return_value = MagicMock()
+    mock_sqlite.return_value = MagicMock()
+
+    mock_vector_store = MagicMock()
+    mem_a = MagicMock()
+    mem_a.id = "mem-a"
+    mem_a.payload = {"data": "Alice likes Python", "user_id": "alice"}
+    mem_b = MagicMock()
+    mem_b.id = "mem-b"
+    mem_b.payload = {"data": "Alice works at Acme", "user_id": "alice"}
+    mock_vector_store.list.return_value = ([mem_a, mem_b],)
+    mock_vector_store.get.side_effect = lambda vector_id: {"mem-a": mem_a, "mem-b": mem_b}[vector_id]
+    mock_vector_factory.return_value = mock_vector_store
+
+    mock_entity_store = MagicMock()
+    entity_row = MagicMock()
+    entity_row.id = "entity-alice"
+    entity_row.payload = {
+        "data": "alice",
+        "user_id": "alice",
+        "linked_memory_ids": ["mem-a", "mem-b"],
+    }
+    mock_entity_store.list.return_value = ([entity_row],)
+
+    from mem0.memory.main import AsyncMemory
+    config = MemoryConfig()
+    memory = AsyncMemory(config)
+    memory._entity_store = mock_entity_store
+
+    await memory.delete_all(user_id="alice")
+
+    mock_entity_store.delete.assert_called_once_with(vector_id="entity-alice")
+
+    assert mock_vector_store.delete.call_count == 2

--- a/tests/test_memory.py
+++ b/tests/test_memory.py
@@ -1258,50 +1258,52 @@ class TestPreserveCustomMetadata:
         assert payload["user_id"] == "user_1"
 
 
-@pytest.mark.asyncio
-@patch('mem0.utils.factory.EmbedderFactory.create')
-@patch('mem0.utils.factory.VectorStoreFactory.create')
-@patch('mem0.utils.factory.LlmFactory.create')
-@patch('mem0.memory.storage.SQLiteManager')
-async def test_async_delete_all_bulk_clears_entity_store(mock_sqlite, mock_llm_factory, mock_vector_factory, mock_embedder_factory):
-    """
-    Verify that async delete_all bulk-clears entity records before running
-    concurrent memory deletes, preventing the race condition where concurrent
-    _remove_memory_from_entity_store calls read stale snapshots of
-    linked_memory_ids and overwrite each other's writes.
-    """
-    mock_embedder_factory.return_value = MagicMock()
-    mock_llm_factory.return_value = MagicMock()
-    mock_sqlite.return_value = MagicMock()
+class TestAsyncDeleteAllEntityRace:
+    """Tests for async delete_all entity store race condition fix."""
 
-    mock_vector_store = MagicMock()
-    mem_a = MagicMock()
-    mem_a.id = "mem-a"
-    mem_a.payload = {"data": "Alice likes Python", "user_id": "alice"}
-    mem_b = MagicMock()
-    mem_b.id = "mem-b"
-    mem_b.payload = {"data": "Alice works at Acme", "user_id": "alice"}
-    mock_vector_store.list.return_value = ([mem_a, mem_b],)
-    mock_vector_store.get.side_effect = lambda vector_id: {"mem-a": mem_a, "mem-b": mem_b}[vector_id]
-    mock_vector_factory.return_value = mock_vector_store
+    @pytest.mark.asyncio
+    @patch('mem0.utils.factory.EmbedderFactory.create')
+    @patch('mem0.utils.factory.VectorStoreFactory.create')
+    @patch('mem0.utils.factory.LlmFactory.create')
+    @patch('mem0.memory.storage.SQLiteManager')
+    async def test_async_delete_all_bulk_clears_entity_store(self, mock_sqlite, mock_llm_factory, mock_vector_factory, mock_embedder_factory):
+        """
+        Verify that async delete_all bulk-clears entity records after
+        concurrent memory deletes complete, preventing both the
+        read-modify-write race and entity orphaning on partial failures.
+        """
+        mock_embedder_factory.return_value = MagicMock()
+        mock_llm_factory.return_value = MagicMock()
+        mock_sqlite.return_value = MagicMock()
 
-    mock_entity_store = MagicMock()
-    entity_row = MagicMock()
-    entity_row.id = "entity-alice"
-    entity_row.payload = {
-        "data": "alice",
-        "user_id": "alice",
-        "linked_memory_ids": ["mem-a", "mem-b"],
-    }
-    mock_entity_store.list.return_value = ([entity_row],)
+        mock_vector_store = MagicMock()
+        mem_a = MagicMock()
+        mem_a.id = "mem-a"
+        mem_a.payload = {"data": "Alice likes Python", "user_id": "alice"}
+        mem_b = MagicMock()
+        mem_b.id = "mem-b"
+        mem_b.payload = {"data": "Alice works at Acme", "user_id": "alice"}
+        mock_vector_store.list.return_value = ([mem_a, mem_b],)
+        mock_vector_store.get.side_effect = lambda vector_id: {"mem-a": mem_a, "mem-b": mem_b}[vector_id]
+        mock_vector_factory.return_value = mock_vector_store
 
-    from mem0.memory.main import AsyncMemory
-    config = MemoryConfig()
-    memory = AsyncMemory(config)
-    memory._entity_store = mock_entity_store
+        mock_entity_store = MagicMock()
+        entity_row = MagicMock()
+        entity_row.id = "entity-alice"
+        entity_row.payload = {
+            "data": "alice",
+            "user_id": "alice",
+            "linked_memory_ids": ["mem-a", "mem-b"],
+        }
+        mock_entity_store.list.return_value = ([entity_row],)
 
-    await memory.delete_all(user_id="alice")
+        from mem0.memory.main import AsyncMemory
+        config = MemoryConfig()
+        memory = AsyncMemory(config)
+        memory._entity_store = mock_entity_store
 
-    mock_entity_store.delete.assert_called_once_with(vector_id="entity-alice")
+        await memory.delete_all(user_id="alice")
 
-    assert mock_vector_store.delete.call_count == 2
+        mock_entity_store.delete.assert_called_once_with(vector_id="entity-alice")
+
+        assert mock_vector_store.delete.call_count == 2

